Mathlib/FieldTheory/PurelyInseparable.lean

```diff
@@ -0,0 +1,579 @@
+    finSepDegree_eq_finrank_of_isSeparable E _, one_mul,
+    finrank_eq_one_iff (F := E) (E := AdjoinRoot f)] at hdeg
+  obtain ⟨x, h⟩ := hdeg ▸ mem_adjoin_simple_self E (AdjoinRoot.root f)
+  replace h := congr(aeval $h (minpoly E (AdjoinRoot.root f)))
+  erw [minpoly.aeval, hf, aeval_algebraMap_apply,
+    ← map_zero (f := algebraMap E (AdjoinRoot f))] at h
+  exact ⟨x, (algebraMap E _).injective h⟩
+
+end IsPurelyInseparable
+
+namespace IntermediateField
+
+instance isPurelyInseparable_bot : IsPurelyInseparable F (⊥ : IntermediateField F E) :=
+  (botEquiv F E).symm.isPurelyInseparable
+
+/-- `F⟮x⟯ / F` is a purely inseparable extension if and only if the mininal polynomial of `x`
+has separable degree one. -/
+theorem isPurelyInseparable_adjoin_simple_iff_natSepDegree_eq_one {x : E} :
+    IsPurelyInseparable F F⟮x⟯ ↔ (minpoly F x).natSepDegree = 1 := by
+  refine ⟨fun h ↦ ?_, fun h ↦ ?_⟩
+  · have := (isPurelyInseparable_iff_natSepDegree_eq_one F _).1 h (AdjoinSimple.gen F x)
+    rwa [minpoly_eq] at this
+  have hx : IsIntegral F x := by_contra fun h' ↦ by
+    simp only [minpoly.eq_zero h', natSepDegree_zero, zero_ne_one] at h
+  haveI := adjoin.finiteDimensional hx
+  rwa [isPurelyInseparable_iff_finSepDegree_eq_one _ _ (Algebra.IsAlgebraic.of_finite F F⟮x⟯),
+    finSepDegree_adjoin_simple_eq_natSepDegree _ _ hx.isAlgebraic]
+
+/-- If `F` is of exponential characteristic `q`, then `F⟮x⟯ / F` is a purely inseparable extension
+if and only if `x ^ (q ^ n)` is contained in `F` for some `n : ℕ`. -/
+theorem isPurelyInseparable_adjoin_simple_iff_mem_pow (q : ℕ) [hF : ExpChar F q] {x : E} :
+    IsPurelyInseparable F F⟮x⟯ ↔ ∃ n : ℕ, x ^ q ^ n ∈ (algebraMap F E).range := by
+  simp_rw [isPurelyInseparable_adjoin_simple_iff_natSepDegree_eq_one,
+    minpoly.natSepDegree_eq_one_iff_mem_pow q]
+
+-- Marked as private since it's a special case of `isPurelyInseparable_adjoin_iff_mem_pow.2`.
+private theorem isPurelyInseparable_adjoin_finset_of_mem_pow (q : ℕ) [hF : ExpChar F q]
+    (S : Finset E) (h : ∀ x ∈ S, ∃ n : ℕ, x ^ q ^ n ∈ (algebraMap F E).range) :
+    IsPurelyInseparable F (adjoin F (S : Set E)) := by
+  refine induction_on_adjoin_finset S (IsPurelyInseparable F ·) inferInstance fun L x hx hL ↦ ?_
+  -- I don't know why, but `simp only at hL` is not regarded as an instance
+  change IsPurelyInseparable F L at hL
+  obtain ⟨n, y, h⟩ := h x hx
+  haveI := expChar_of_injective_algebraMap (algebraMap F L).injective q
+  replace h := (isPurelyInseparable_adjoin_simple_iff_mem_pow L E q).2 ⟨n, (algebraMap F L) y, h⟩
+  exact IsPurelyInseparable.trans F L L⟮x⟯
+
+/-- If `F` is of exponential characteristic `q`, then `F(S) / F` is a purely inseparable extension
+if and only if for any `x ∈ S`, `x ^ (q ^ n)` is contained in `F` for some `n : ℕ`. -/
+theorem isPurelyInseparable_adjoin_iff_mem_pow (q : ℕ) [hF : ExpChar F q] {S : Set E} :
+    IsPurelyInseparable F (adjoin F S) ↔ ∀ x ∈ S, ∃ n : ℕ, x ^ q ^ n ∈ (algebraMap F E).range := by
```

Comment:
Do we have `IsPurelyInseparable F E ↔ ∀ x : E, ∃ n, x ^ q ^ n ∈ (algebraMap F E).range`?

Reply:
Yes, it's `isPurelyInseparable_iff_mem_pow`.